File: java/src/player/property/Team.java
package player.property;

import java.awt.Color;
import java.util.Arrays;

import misc.Debug;
import misc.compress.*;

public enum Team implements Compressable
{
	TEAM0(0, "None", "666666"),
	TEAM1(1, "Red", "ff0000"),
	TEAM2(2, "Blue", "0000ff"),
	TEAM3(3, "Green", "00ff00"),
	TEAM4(4, "Yellow", "ffff00");

	private int id;
	private String name;
	private int colorID;

	private Team(int id, String name, String colorString)
	{
		this.id = id;
		this.name = name;
		this.colorID = Integer.parseInt(colorString, 16);
	}

	public boolean isTeamFriendly(Team team)
	{
		if (this == TEAM0 || team == TEAM0)
			return false;
		return this == team;
	}

	public static Team getByID(int id)
	{
		for (Team team : Team.values())
		{
			if (team.getID() == id)
			{
				return team;
			}
		}
		Debug.error("team id " + id + " in Team.getByID");
		return null;
	}

	@Override public byte getCID() { return Compressor.TEAM_CID; }
	@Override public byte[] compress()
	{
		return new byte[]{(byte) Arrays.asList(Team.values()).indexOf(this)};
	}

	public static Team cut(CompressBuffer buffer)
	{
		byte b = buffer.cutByte();
		return Team.values()[(int) b];
	}


	public int getID() { return id; }
	public String getName() { return name; }
	public Color getColor() { return new Color(getColorID()); }
	public int getColorID() { return colorID; }
}

File: cpp/src/player/property/Team.cpp
#include "Team.hpp"

#include <core/Main.hpp>
#include <misc/Debug.hpp>
#include <misc/Converter.hpp>

std::vector<Team*> Team::teams;

Team::Team(char id, const std::string& name, const std::string& color)
	: id(id), name(name), colorID(Converter::colorStringToInt(color)) // TODO define color id
{}

void Team::init()
{
	if (teams.size() != 0) Debug::warn("Team::init() was already called");
	teams.push_back(new Team(0, "None", "666666"));
	teams.push_back(new Team(1, "Red", "ff0000"));
	teams.push_back(new Team(2, "Blue", "0000ff"));
	teams.push_back(new Team(3, "Green", "00ff00"));
	teams.push_back(new Team(4, "Yellow", "ffff00"));
}

void Team::uninit()
{
	deleteAndClearVector(teams);
}

Team* Team::get(int id)
{
	if (teams.size() == 0) Debug::warn("Team::get(): Teams not yet initialized");
	return teams[id];
}

unsigned int Team::getAmount()
{
	if (teams.size() == 0) Debug::warn("Team::getAmount(): == 0");
	return teams.size();
}

std::string Team::getName() const
{
	return name;
}

char Team::getID() const
{
	return id;
}

int Team::getColorID() const
{
	return colorID;
}

bool Team::isFriendlyTeam(Team* team) const
{
	return (getID() != 0) && (getID() == team->getID());
}
